test/test_sde.py

```diff
@@ -0,0 +1,245 @@
+    yield diffrax.SPaRK, "add", 1.5
+    yield diffrax.GeneralShARK, "add", 1.5
+    yield diffrax.ShARK, "add", 1.5
+    yield diffrax.SRA1, "add", 1.5
+    yield diffrax.SEA, "add", 1.0
+
+
+# TODO: For solvers of high order, comparing to Euler or Heun is not good,
+# because they are waaaay worse than for example ShARK. ShARK is more precise
+# at dt=2**-4 than Euler is at dt=2**-14 (and it takes forever to run at such
+# a small dt). Hence , the order of convergence of ShARK seems to plateau at
+# discretisations finer than 2**-4.
+# I propose the following:
+# We can sparate this test into two. First we determine how fast the solver
+# converges to its own limit (i.e. using itself as reference), and then
+# check whether that limit is the same as the Euler/Heun limit.
+# For the second, I would like to make a separate check, where the "correct"
+# solution is computed only once and then all solvers are compared to it.
+@pytest.mark.parametrize("solver_ctr,noise,theoretical_order", _solvers_and_orders())
+def test_sde_strong_order_new(
+    solver_ctr, noise: Literal["any", "com", "add"], theoretical_order
+):
+    bmkey = jr.PRNGKey(5678)
+    sde_key = jr.PRNGKey(11)
+    num_samples = 100
+    bmkeys = jr.split(bmkey, num=num_samples)
+    t0 = 0.3
+    t1 = 5.3
+
+    if noise == "add":
+        sde = get_time_sde(t0, t1, jnp.float64, sde_key, noise_dim=7)
+    else:
+        if noise == "com":
+            noise_dim = 1
+        elif noise == "any":
+            noise_dim = 5
+        else:
+            assert False
+        sde = get_mlp_sde(t0, t1, jnp.float64, sde_key, noise_dim=noise_dim)
+
+    ref_solver = solver_ctr()
+    level_coarse, level_fine = 4, 10
+
+    # We specify the times to which we step in way that each level contains all the
+    # steps of the previous level. This is so that we can compare the solutions at
+    # all the times in saveat, and not just at the end time.
+    def get_dt_step_controller(level):
+        step_ts = jnp.linspace(t0, t1, 2**level + 1, endpoint=True)
+        return None, diffrax.StepTo(ts=step_ts)
+
+    saveat = diffrax.SaveAt(ts=jnp.linspace(t0, t1, 2**level_coarse + 1, endpoint=True))
+
+    hs, errors, order = simple_sde_order(
+        bmkeys,
+        sde,
+        solver_ctr(),
+        ref_solver,
+        (level_coarse, level_fine),
+        get_dt_step_controller,
+        saveat,
+        bm_tol=2**-14,
+    )
+    # The upper bound needs to be 0.25, otherwise we fail.
+    # This still preserves a 0.05 buffer between the intervals
+    # corresponding to the different orders.
+    print(order)
+    assert -0.2 < order - theoretical_order < 0.25
+
+
+# Make variables to store the correct solutions in.
+# This is to avoid recomputing the correct solutions for every solver.
+solutions = {
+    "Ito": {
+        "any": None,
+        "com": None,
+        "add": None,
+    },
+    "Stratonovich": {
+        "any": None,
+        "com": None,
+        "add": None,
+    },
+}
```

Comment:
Heads-up that this kind of caching is usually best done with a pytest fixture -- in this case probably one that is module-scoped. (So that it is torn down once all tests in the module are completed.)

In practice in Diffrax we actually run every module in a separate process (see `__main__.py`), so the distinction doesn't matter here and I'm not asking you to change it; I'm just making a comment as a teaching point.

Reply:
That's very good to know, thanks!